File: SeongJae/week3/BOJ_2178.py
import sys
from collections import deque
def bfs(arr,visited):
   queue=deque()
   queue.append((0,0,1))
   visited[0][0]=True
   while queue:
      x,y,count=queue.popleft()
      for i in range(4):
         ax=x+dx[i]
         ay=y+dy[i]
         if 0<=ax<m and 0<=ay<n and not visited[ay][ax] and arr[ay][ax]==1:
            queue.append((ax,ay,count+1))
            visited[ay][ax]=True
            if ay==n-1 and ax==m-1:
               return count+1

input=sys.stdin.readline
n,m=map(int,input().split())
arr=[list(map(int,input().rstrip())) for i in range(n)]
visited=[[False for i in range(m)] for j in range(n)]
dx=[1,-1,0,0]
dy=[0,0,1,-1]
print(bfs(arr,visited))

File: hyobin/week3/BOJ_2178.java
import java.io.*;
import java.util.*;

public class BOJ_2178 {
    static int n;
    static int m;
    static int[][] arr;
    static int[] dx = {1,-1,0,0};
    static int[] dy = {0,0,1,-1};
    static int[][] visited;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        visited = new int[n][m];
        for (int i =0; i<n; i++){
            String[] str = br.readLine().split("");
            for(int j = 0 ;j<m;j++ ){
                arr[i][j] = Integer.parseInt(str[j]);
            }
        }
        bfs();
        System.out.println(visited[n-1][m-1]);
    }
    static void bfs(){
        Queue<int[]> queue = new ArrayDeque<>();
        queue.add(new int[] {0,0});
        visited[0][0] = 1;

        while (!queue.isEmpty()){
            int[] temp = queue.poll();
            int x = temp[0];
            int y = temp[1];
            for (int i =0; i <4; i++){
                int nx = x + dx[i];
                int ny = y+ dy[i];

                if (0<=nx && nx<n && 0<= ny && ny<m){
                    if (arr[nx][ny]==1 && ( visited[nx][ny] == 0 | visited[nx][ny] > visited[x][y]+1 )){
                        visited[nx][ny] =  visited[x][y]+1;
                        queue.add(new int[] {nx,ny});
                    }
                }
            }
        }

    }

}
